cmd/kubeadm/app/util/runtime/runtime.go

```diff
@@ -179,3 +181,52 @@ func (runtime *DockerRuntime) ImageExists(image string) (bool, error) {
 	err := runtime.exec.Command("docker", "inspect", image).Run()
 	return err == nil, nil
 }
+
+// isExistingSocket checks if path exists and is domain socket
+func isExistingSocket(path string) bool {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+
+	return fileInfo.Mode()&os.ModeSocket != 0
+}
+
+// detectCRISocketImpl is separated out only for test purposes, DON'T call it directly, use DetectCRISocket instead
+func detectCRISocketImpl(isSocket func(string) bool) (string, error) {
+	foundCRISockets := []string{}
+	knownCRISockets := []string{
+		// /var/run/dockershim.sock is deliberately omitted from here, we have a separate Docker check below
+		"/run/containerd/containerd.sock",
+		"/var/run/crio/crio.sock",
+	}
+
+	const dockerSocket = "/var/run/docker.sock" // The Docker socket is not CRI compatible
+	if isSocket(dockerSocket) {
+		foundCRISockets = append(foundCRISockets, kubeadmapiv1alpha3.DefaultDockerCRISocket)
+	}
+
+	for _, socket := range knownCRISockets {
+		if isSocket(socket) {
+			foundCRISockets = append(foundCRISockets, socket)
+		}
+	}
+
+	switch len(foundCRISockets) {
+	case 0:
+		return "", goerrors.New("No CRI socket detected, please use the --cri-socket option")
+	case 1:
+		return foundCRISockets[0], nil
+	default:
```

Comment:
i wonder it would be a better or worse experience to print a warning that multiple sockets were detected and kubeadm is continuing with the first one it found? My thought is that this would allow more users to get started without having to specify a specific socket to use, but still be aware there may be a problem.

Reply:
Every possible solution, from UX perspective, is a double edged sword.

On one hand, if we bail with error (the current version) the user will have to re-run the command with correct `--cri-socket` passed in. This may be annoying to some users and it's still not guaranteed, that they will provide the correct CRI.

On other hand, if we warn the user and just pick up the first one, then the user may ignore or not see the warning and end up in a mess (requiring reset).

For me the first option is more viable in the case of multiple CRIs and therefore I picked that solution.